File: dsa-problems/Strings/solutions/javascript/STR-014-shortest-covering-window-set.js
function shortestCoveringWindow(arr, T) {
  if (!arr || arr.length === 0 || !T || T.size === 0) {
    return [0, []];
  }

  const required = new Map();
  for (let s of T) {
    required.set(s, 1);
  }

  const windowCounts = new Map();
  let left = 0,
    formed = 0;
  let minLen = Infinity;
  let resultLeft = 0,
    resultRight = 0;

  for (let right = 0; right < arr.length; right++) {
    const s = arr[right];
    windowCounts.set(s, (windowCounts.get(s) || 0) + 1);

    if (required.has(s) && windowCounts.get(s) === 1) {
      formed++;
    }

    while (formed === T.size && left <= right) {
      if (right - left + 1 < minLen) {
        minLen = right - left + 1;
        resultLeft = left;
        resultRight = right;
      }

      const leftS = arr[left];
      windowCounts.set(leftS, windowCounts.get(leftS) - 1);
      if (required.has(leftS) && windowCounts.get(leftS) === 0) {
        formed--;
      }

      left++;
    }
  }

  if (minLen === Infinity) {
    return [0, []];
  }

  return [minLen, arr.slice(resultLeft, resultRight + 1)];
}
































const readline = require('readline');
const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
let tokens = [];
rl.on('line', (line) => { tokens.push(...line.trim().split(/\s+/)); });
rl.on('close', () => {
    if(tokens.length===0) return;
    let ptr = 0;
    const arr_n = parseInt(tokens[ptr++]);
    const arr = [];
    for(let i=0; i<arr_n; i++) arr.push(tokens[ptr++]);
    const T_n = parseInt(tokens[ptr++]);
    const T = new Set();
    for(let i=0; i<T_n; i++) T.add(tokens[ptr++]);
    const res = shortestCoveringWindow(arr, T);
    console.log(res[0]);
    if(res[1].length > 0) { for(const s of res[1]) console.log(s); } else { console.log('NONE'); }
});

File: dsa-problems/Strings/solutions/python/STR-014-shortest-covering-window-set.py
def shortest_covering_window(arr: list[str], T: set[str]) -> tuple[int, list[str]]:
    if not arr or not T:
        return (0, [])

    required = {s: 1 for s in T}
    window_counts = {}

    left = 0
    formed = 0  # Unique T elements covered
    min_len = float('inf')
    result_left, result_right = 0, 0

    for right in range(len(arr)):
        char = arr[right]
        window_counts[char] = window_counts.get(char, 0) + 1

        # Check if this char is in T and newly covered
        if char in required and window_counts[char] == 1:
            formed += 1

        # Contract window while all elements covered
        while formed == len(T) and left <= right:
            # Update result if better
            if right - left + 1 < min_len:
                min_len = right - left + 1
                result_left, result_right = left, right

            # Remove leftmost element
            left_char = arr[left]
            window_counts[left_char] -= 1
            if left_char in required and window_counts[left_char] == 0:
                formed -= 1

            left += 1

    if min_len == float('inf'):
        return (0, [])

    return (min_len, arr[result_left:result_right + 1])


def main():
    import sys


    # Read input
    input_data = sys.stdin.read().strip()
    if not input_data:
        print(0)
        return
        
    parts = input_data.split()
    if not parts:
        return

    iterator = iter(parts)
    try:
        # Read Array
        N = int(next(iterator))
        arr = []
        for _ in range(N):
            arr.append(next(iterator))
            
        # Read Set T
        K = int(next(iterator))
        T = set()
        for _ in range(K):
            T.add(next(iterator))
            
        length, window = shortest_covering_window(arr, T)
        
        print(length)
        for item in window:
            print(item)
            
    except StopIteration:
        pass
    except ValueError:
        pass

if __name__ == "__main__":
    main()
